metron-platform/metron-parsers/3rdPartyParser.md

```diff
@@ -0,0 +1,306 @@
+import org.apache.metron.parsers.BasicParser;
+import org.json.simple.JSONObject;
+
+import java.util.List;
+import java.util.Map;
+
+public class SimpleParser extends BasicParser {
+  @Override
+  public void init() {
+
+  }
+
+  @Override
+  public List<JSONObject> parse(byte[] bytes) {
+    String input = new String(bytes);
+    Iterable<String> it = Splitter.on(",").split(input);
+    JSONObject ret = new JSONObject();
+    ret.put("original_string", input);
+    ret.put("timestamp", System.currentTimeMillis());
+    ret.put("first", Iterables.getFirst(it, "missing"));
+    ret.put("last", Iterables.getLast(it, "missing"));
+    return ImmutableList.of(ret);
+  }
+
+  @Override
+  public void configure(Map<String, Object> map) {
+
+  }
+}
+```
+* Compile the parser via `mvn clean package` in `extra_parsers`
+
+This will create a jar containing your parser and its dependencies (sans Metron dependencies) in `extra-parsers/target/extra-parsers-1.0-SNAPSHOT-uber.jar`
+
+# Deploying Your Custom Parser
+
+In order to deploy your newly built custom parser, you would place the jar file above in the `$METRON_HOME/parser_contrib` directory on the Metron host (i.e. any host you would start parsers from or, alternatively, where the Metron REST is hosted).
+
+## Example
+
+Let's work through deploying the example above.
+
+### Preliminaries
+
+We assume that the following environment variables are set:
+* `METRON_HOME` - the home directory for metron
+* `ZOOKEEPER` - The zookeeper quorum (comma separated with port specified: e.g. `node1:2181` for full-dev)
+* `BROKERLIST` - The Kafka broker list (comma separated with port specified: e.g. `node1:6667` for full-dev)
+* `ES_HOST` - The elasticsearch master (and port) e.g. `node1:9200` for full-dev.
+
+Also, this does not assume that you are using a kerberized cluster.  If you are, then the parser start command will adjust slightly to include the security protocol.
+
+### Copy the jar file up
+
+Copy the jar file located in `extra-parsers/target/extra-parsers-1.0-SNAPSHOT-uber.jar` to `$METRON_HOME/parser_contrib` and ensure the permissions are such that the `metron` user can read and execute.
+
+### Restart the REST service in Ambari
+
+In order for new parsers to be picked up, the REST service must be restarted.  You can do that from within Ambari by restarting the `Metron REST` service.
+
+### Create a Kafka Topic
+
+Create a kafka topic, let's call it `test` via:
+`/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper $ZOOKEEPER --create --topic test --partitions 1 --replication-factor 1`
+
+Note, in a real deployment, that topic would be named something more descriptive and would have replication factor and partitions set to something less trivial.
+
+### Configure Test Parser
+
+Create the a file called `$METRON_HOME/config/zookeeper/parsers/test.json` with the following content:
+```
+{
+  "parserClassName":"com.thirdparty.SimpleParser",
+  "sensorTopic":"test"
+}
+```
+
+### Start Parser
```

Comment:
Need to push config to Zookeeper before starting the parser.

```
Now push the config to Zookeeper with the following command:
$METRON_HOME/bin/zk_load_configs.sh -m PUSH -i $METRON_HOME/config/zookeeper/ -z $ZOOKEEPER

```

Reply:
WHoops, missed that one.  Thanks, mike!